ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+			field.Error: err,
+		})
+	}
+
+	cpu := runtime.NumCPU() * 1024
+
+	return int64(cpu), mem
+}
+
+func validateRegisteredAttributes(expectedAttributes, actualAttributes []*ecsmodel.Attribute) error {
+	var err error
+	expectedAttributesMap := attributesToMap(expectedAttributes)
+	actualAttributesMap := attributesToMap(actualAttributes)
+	missingAttributes, err := findMissingAttributes(expectedAttributesMap, actualAttributesMap)
+	if err != nil {
+		msg := strings.Join(missingAttributes, ",")
+		logger.Error("Error registering attributes", logger.Fields{
+			field.Error:         err,
+			"missingAttributes": msg,
+		})
+	}
+	return err
+}
+
+func (client *ecsClient) getAdditionalAttributes() []*ecsmodel.Attribute {
+	attrs := []*ecsmodel.Attribute{
+		{
+			Name:  aws.String(osTypeAttrName),
+			Value: aws.String(client.configAccessor.OSType()),
+		},
+		{
+			Name:  aws.String(osFamilyAttrName),
+			Value: aws.String(client.configAccessor.OSFamily()),
+		},
+	}
+	// Send CPU arch attribute directly when running on external capacity. When running on EC2 or Fargate launch type,
+	// this is not needed since the CPU arch is reported via instance identity document in those cases.
+	if client.configAccessor.External() {
+		attrs = append(attrs, &ecsmodel.Attribute{
+			Name:  aws.String(cpuArchAttrName),
+			Value: aws.String(getCPUArch()),
+		})
+	}
+	return attrs
+}
+
+func (client *ecsClient) getOutpostAttribute(outpostARN string) []*ecsmodel.Attribute {
+	if len(outpostARN) > 0 {
+		return []*ecsmodel.Attribute{
+			{
+				Name:  aws.String("ecs.outpost-arn"),
+				Value: aws.String(outpostARN),
+			},
+		}
+	}
+	return []*ecsmodel.Attribute{}
+}
+
+func (client *ecsClient) getCustomAttributes() []*ecsmodel.Attribute {
+	var attributes []*ecsmodel.Attribute
+	for attribute, value := range client.configAccessor.InstanceAttributes() {
+		attributes = append(attributes, &ecsmodel.Attribute{
+			Name:  aws.String(attribute),
+			Value: aws.String(value),
+		})
+	}
+	return attributes
+}
+
+func (client *ecsClient) SubmitTaskStateChange(change ecs.TaskStateChange) error {
+	if change.Attachment != nil && client.stscAttachmentCustomRetryBackoff != nil {
+		retryFunc := func() error {
+			err := client.submitTaskStateChange(change)
+			if err == nil {
+				return nil
+			}
+			return submitStateCustomRetriableError(err)
+		}
+		return client.stscAttachmentCustomRetryBackoff(retryFunc)
+	}
+	return client.submitTaskStateChange(change)
+}
+
+func (client *ecsClient) submitTaskStateChange(change ecs.TaskStateChange) error {
+	if change.Attachment != nil {
+		// Confirm attachment by submitting attachment state change via SubmitTaskStateChange API (specifically in
+		// the input's Attachments field).
+		var attachments []*ecsmodel.AttachmentStateChange
+		eniStatus := change.Attachment.Status.String()
+		attachments = []*ecsmodel.AttachmentStateChange{
+			{
+				AttachmentArn: aws.String(change.Attachment.AttachmentARN),
```

Comment:
nit: If this is the `ni` type I think it is, it seems we prefer to access with the mutex locked `GetAttachmentArn()`?

https://github.com/aws/amazon-ecs-agent/blob/master/ecs-agent/netlib/model/networkinterface/eniattachment.go#L160

Reply:
Synced offline, this is a non-blocking nit and we will keep as is since it mirrors current existing code/behavior.